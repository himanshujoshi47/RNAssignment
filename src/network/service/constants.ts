export enum CachePolicy {
  ALWAYS_IGNORE_CACHE,
  RETURN_CACHE_ELSE_LOAD,
  INVALIDATE_CACHE_AND_LOAD,
}

export interface IService {
  fetch: (endpoint: string, cachePolicy?: CachePolicy) => Promise<any>;
}

export const Endpoint = {
  GET_FOOD_LIST: 'https://api.jsonbin.io/b/5fce7e1e2946d2126fff85f0',
};
