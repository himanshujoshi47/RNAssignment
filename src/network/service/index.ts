import Service from './service';
import {CachePolicy, IService, Endpoint} from './constants';

export type {CachePolicy, IService};
export {Endpoint};
export {Service};
