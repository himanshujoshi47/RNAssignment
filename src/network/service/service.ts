import {
  GetLastPathName,
  GetLocalItem,
  RemoveLocalItem,
  SetLocalItem,
} from '../LocalStorage/LocalStorage';
import {CachePolicy, IService} from './constants';

class Service implements IService {
  private static _instance = new Service();
  private constructor() {}
  static get instance() {
    return this._instance;
  }

  fetch = async (
    endpoint: string,
    cachePolicy: CachePolicy = CachePolicy.RETURN_CACHE_ELSE_LOAD,
  ): Promise<any> => {
    const pathName = GetLastPathName(endpoint);

    switch (cachePolicy) {
      case CachePolicy.ALWAYS_IGNORE_CACHE: {
        return this.getNetworkResponse(endpoint);
      }
      case CachePolicy.RETURN_CACHE_ELSE_LOAD: {
        let cacheData = await GetLocalItem(pathName);
        if (cacheData) {
          return cacheData;
        } else {
          let networkResponse = await this.getNetworkResponse(endpoint);
          SetLocalItem(pathName, networkResponse);
          return networkResponse;
        }
      }
      case CachePolicy.INVALIDATE_CACHE_AND_LOAD: {
        RemoveLocalItem(pathName);
        let networkResponse = await this.getNetworkResponse(endpoint);
        SetLocalItem(pathName, networkResponse);
        return networkResponse;
      }
    }
  };

  private getNetworkResponse = async (endpoint: string): Promise<any> => {
    try {
      let response = await fetch(endpoint);
      return response.json();
    } catch (error) {
      console.error(error);
    }
  };
}

export default Service.instance;
