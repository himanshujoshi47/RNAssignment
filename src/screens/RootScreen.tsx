import React, {useCallback, useState} from 'react';
import {
  SafeAreaView,
  View,
  StyleSheet,
  ActivityIndicator,
  Text,
} from 'react-native';
import {Button} from 'react-native-elements';
import {Service, Endpoint, IService} from '../network/service';
import R from '../res';
import FoodListModal from './FoodListModal';
import {FoodListModel} from '../network/models/FoodListModel';

interface FoodListData {
  showFoodList: boolean;
  errorMessage: string | null;
  data: FoodListModel | null;
}

const PlayerScreen: React.FC<{}> = () => {
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [foodList, setFoodList] = useState<FoodListData>({
    showFoodList: false,
    errorMessage: null,
    data: null,
  });

  const getService = (): IService => {
    return Service;
  };

  const getFoodList = async () => {
    setIsLoading(true);
    let res =
      foodList.data ?? (await getService().fetch(Endpoint.GET_FOOD_LIST));
    setIsLoading(false);
    if (res) {
      setFoodList((prevState: FoodListData) => {
        return {
          ...prevState,
          showFoodList: true,
          data: res,
        };
      });
    } else {
      setFoodList((prevState: FoodListData) => {
        return {
          ...prevState,
          showFoodList: false,
          error: 'Sorry! Cannot fetch food list this time',
        };
      });
    }
  };

  const didCloseFoodListModal = useCallback(() => {
    setFoodList((prevState: FoodListData) => {
      return {
        ...prevState,
        showFoodList: false,
      };
    });
  }, []);

  return (
    <SafeAreaView style={localStyles.safeArea}>
      {foodList.errorMessage && <Text>{foodList.errorMessage}</Text>}
      <View style={localStyles.container}>
        <ActivityIndicator
          testID={R.id.FOOD_LIST_LOADER}
          animating={isLoading}
          hidesWhenStopped={true}
          size="large"
        />
        <Button
          title={R.strings.APPROVED_FOOD_LIST}
          containerStyle={localStyles.containerStyle}
          testID={R.id.APPROVED_FOOD_LIST_BTN}
          onPress={() => getFoodList()}
        />
      </View>
      <FoodListModal
        isVisible={foodList.showFoodList}
        data={foodList.data}
        didCloseFoodListModal={didCloseFoodListModal}
      />
    </SafeAreaView>
  );
};

const localStyles = StyleSheet.create({
  safeArea: {
    flex: 1,
    backgroundColor: R.colors.background.silver,
  },
  container: {
    width: '100%',
    height: '100%',
    justifyContent: 'flex-end',
    alignItems: 'center',
    padding: R.dimensions.dimen_24,
  },
  containerStyle: {
    width: '100%',
    justifyContent: 'center',
  },
});

export default PlayerScreen;
