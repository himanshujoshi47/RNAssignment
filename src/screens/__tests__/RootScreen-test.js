import React from 'react';
import renderer, {act} from 'react-test-renderer';
import {Service} from '../../network/service';
import R from '../../res';
import RootScreen from '../RootScreen';

jest.useFakeTimers();

describe('renders and behave correctly', () => {
  const rootScreen = renderer.create(<RootScreen />);
  const rootScreenInst = rootScreen.root;

  it('renders correctly', () => {
    expect(rootScreen).toMatchSnapshot();
  });

  it('should have required elements', () => {
    expect(
      rootScreenInst.findByProps({testID: R.id.APPROVED_FOOD_LIST_BTN}),
    ).toBeTruthy();
  });

  it('should fetch the data on button press', () => {
    act(() => {
      rootScreenInst
        .findByProps({testID: R.id.APPROVED_FOOD_LIST_BTN})
        .props.onPress();
    });
    expect(Service.fetch).toBeCalled();
  });
});
