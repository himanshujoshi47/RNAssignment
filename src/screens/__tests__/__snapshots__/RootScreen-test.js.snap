// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and behave correctly renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#ECECEC",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": "100%",
        "justifyContent": "flex-end",
        "padding": 24,
        "width": "100%",
      }
    }
  >
    <ActivityIndicator
      animating={false}
      color="#999999"
      hidesWhenStopped={true}
      size="large"
      testID="food-list-loader"
    />
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    visible={false}
  >
    <RNCSafeAreaView
      style={
        Object {
          "backgroundColor": "#ECECEC",
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "height": "100%",
            "padding": 24,
            "width": "100%",
          }
        }
      >
        <RCTScrollView
          ItemSeparatorComponent={[Function]}
          ListHeaderComponent={
            <Memo(FoodListHeader)
              didCloseFoodListModal={[Function]}
              performSearch={[Function]}
            />
          }
          data={Array []}
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          onContentSizeChange={[Function]}
          onEndReachedThreshold={2}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          showsVerticalScrollIndicator={false}
          stickyHeaderIndices={Array []}
          updateCellsBatchingPeriod={50}
          viewabilityConfigCallbackPairs={Array []}
          windowSize={21}
        >
          <View>
            <View
              onLayout={[Function]}
            >
              <View>
                <Text
                  style={
                    Object {
                      "color": "#222222",
                      "fontSize": 24,
                      "fontWeight": "500",
                      "marginVertical": 24,
                    }
                  }
                  testID="approved-food-list-title"
                >
                  Approved Foods List
                </Text>
                <View
                  style={
                    Object {
                      "backgroundColor": "#FFFFFF4D",
                      "borderRadius": 4,
                      "flexDirection": "row",
                      "marginBottom": 16,
                      "padding": 16,
                    }
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    autoCapitalize="none"
                    autoCorrect={false}
                    onChangeText={[Function]}
                    placeholder="Try searching fat, sauces names..."
                    placeholderTextColor="#CCCCCC"
                    rejectResponderTermination={true}
                    style={
                      Object {
                        "color": "#222222",
                        "flex": 1,
                        "fontSize": 18,
                        "fontWeight": "500",
                        "marginLeft": 8,
                      }
                    }
                    testID="food-list-search-input"
                    underlineColorAndroid="transparent"
                    value=""
                  />
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </RNCSafeAreaView>
  </Modal>
</RCTSafeAreaView>
`;
